### ISSUE TITLE
YAML::XS trips up on tied hashes (like loading to Readonly::Hash hashes)

### ISSUE THREAD

**Issue by MarkHedgesLNE** [ingydotnet/yaml-libyaml-pm#7]
-----Original Message-----
From: s@sankorobinson.com [mailto:s@sankorobinson.com] On Behalf Of Sanko Robinson
Sent: Friday, July 25, 2014 4:25 PM
To: Mark Hedges
Subject: Re: problem with Readonly

Hi!

You've done nothing wrong; YAML::XS::Dump() trips up on tied hashes.
Really strange that no one ran into this before so I tested with
Tie::IxHash to verify it wasn't Readonly specific and the bare bones
demo below also results in the same:

``` perl
use strict;
use warnings;
use YAML::XS;

my %foo = (bar => { ert => 1, yop => 1 }, foo => [qw[baz qux noz]] );

warn "foo:\n".Dump(\%foo);

tie my %bar, 'NewStdHash', %foo;

warn "bar:\n".Dump(\%bar);


package NewStdHash;
require Tie::Hash;
our @ISA = qw[Tie::StdHash];
sub TIEHASH { my $class = shift; return bless {@_}, $class }
sub FIRSTKEY { my $a = scalar keys %{$_[0]}; each %{$_[0]} }
sub NEXTKEY { each %{$_[0]} }
sub FETCH { $_[0]->{$_[1]} }
```

Again, I haven't dug into it yet but you're welcome to use the above
example in a bug report for YAML::XS. The tracker is here:
## https://github.com/ingydotnet/yaml-libyaml-pm/issues

Sanko Robinson
sanko@cpan.org

On Fri, Jul 25, 2014 at 5:58 PM, Mark Hedges
Mark.Hedges@ticketmaster.com wrote:

> BTW this is Readonly v. 2.00.
> 
> Thanks.
> 
> Mark
> 
> -----Original Message-----
> From: Mark Hedges
> Sent: Friday, July 25, 2014 2:57 PM
> To: sanko@cpan.org
> Cc: Mark Hedges
> Subject: problem with Readonly
> 
> What am I doing wrong here?  It seems like this worked before just a few minutes ago.  This is Perl 5.20.0.  Thanks.  -Mark

```
sre@tst2.ntf.cftdev1 /tmp
2014-07-25 14:55:17 Fri  $ ./foo.pl
foo:

---
bar:
  ert: 1
  yop: 1
foo:
- baz
- qux
- noz
bar:

---
bar: 'HASH(0x1ce4ca80)'
foo: 'ARRAY(0x1ce4c9a8)'

sre@tst2.ntf.cftdev1 /tmp
2014-07-25 14:55:20 Fri  $ cat foo.pl
```

``` perl
 #!/ntfhome/local/bin/perl

use strict;
use warnings FATAL => 'all';
use English '-no_match_vars';
use FindBin;
use Readonly;
use YAML::XS;
my %foo = %{ Load(q{

---
foo:
  - baz
  - qux
  - noz
bar:
  yop: 1
  ert: 1
}) };
warn "foo:\n".Dump(\%foo);
Readonly::Hash my %bar => %foo;
warn "bar:\n".Dump(\%bar);
```

**Comment by perlpunk** [comment]
I don't know exactly why it doesn't work, I just see that `SvROK(node)` returns false for a tied hash. Maybe someone with better XS knowledge can help?

**Comment by perlpunk** [comment]
Checkout #101 for a fix

**Comment by perlpunk** [comment]
This is now fixed in v0.83
